File: app/blindshot/engine.ts
/**
 * BlindShot — simulation. No DOM, no React: the render layer reads this and
 * draws it. Mirrors cleanly onto a Roblox server script.
 */

import {
  BALLS, BEDROCK_HP, CELL, GOLDEN_HP, GOLDEN_ROW_MAX, GOLDEN_ROW_MIN,
  GRID_H, GRID_W, KIND_BEDROCK, KIND_GOLDEN, KIND_ROCK, KIND_TOKEN,
  START_CHAMBER, START_COL, START_ROW, TIERS, TOKEN_RARITY,
  makeRng, tierOfRow,
  type BallId, type Stats,
} from './config';

export const CELLS = GRID_W * GRID_H;

export function idx(col: number, row: number): number {
  return row * GRID_W + col;
}

export type World = {
  hp: Int16Array;
  maxHp: Int16Array;
  kind: Uint8Array;
  /** Tier index used for the block's colour — speckled, not always its own tier. */
  hue: Uint8Array;
  /** Per-block brightness jitter, 0..255. */
  shade: Uint8Array;
  seen: Uint8Array;
  poisonT: Float32Array;
  poisonD: Float32Array;
  poisoned: Set<number>;
  goldenIdx: number;
  seed: number;
};

export function inBounds(col: number, row: number): boolean {
  return col >= 0 && col < GRID_W && row >= 0 && row < GRID_H;
}

export function isSolid(w: World, col: number, row: number): boolean {
  if (!inBounds(col, row)) return true;
  return w.hp[idx(col, row)] > 0;
}

// ─── World generation ─────────────────────────────────────────────────────────

export function genWorld(seed: number): World {
  const rng = makeRng(seed);
  const w: World = {
    hp: new Int16Array(CELLS),
    maxHp: new Int16Array(CELLS),
    kind: new Uint8Array(CELLS),
    hue: new Uint8Array(CELLS),
    shade: new Uint8Array(CELLS),
    seen: new Uint8Array(CELLS),
    poisonT: new Float32Array(CELLS),
    poisonD: new Float32Array(CELLS),
    poisoned: new Set<number>(),
    goldenIdx: -1,
    seed,
  };

  for (let row = 0; row < GRID_H; row++) {
    const tier = tierOfRow(row);
    const base = TIERS[tier].hp;
    for (let col = 0; col < GRID_W; col++) {
      const i = idx(col, row);
      w.shade[i] = Math.floor(rng() * 256);

      // Speckle in neighbouring tier colours so seams look like real strata.
      let hue = tier;
      const roll = rng();
      if (roll < 0.09 && tier > 0) hue = tier - 1;
      else if (roll < 0.17 && tier < TIERS.length - 1) hue = tier + 1;
      w.hue[i] = hue;

      const edge = col === 0 || col === GRID_W - 1 || row === 0 || row === GRID_H - 1;
      if (edge) {
        w.kind[i] = KIND_BEDROCK;
        w.hp[i] = BEDROCK_HP;
        w.maxHp[i] = BEDROCK_HP;
        continue;
      }

      const hp = Math.max(1, Math.round(base * (0.75 + rng() * 0.55)));
      w.hp[i] = hp;
      w.maxHp[i] = hp;
      w.kind[i] = rng() * TOKEN_RARITY < 1 ? KIND_TOKEN : KIND_ROCK;
    }
  }

  // A handful of natural voids so the map is not a uniform slab.
  const voids = 26;
  for (let v = 0; v < voids; v++) {
    const cc = 4 + Math.floor(rng() * (GRID_W - 8));
    const cr = 12 + Math.floor(rng() * (GRID_H - 20));
    const rad = 1.5 + rng() * 2.5;
    carve(w, cc, cr, rad);
  }

  // The prize. Always deep, never inside a void, never on the border.
  let gi = -1;
  for (let attempt = 0; attempt < 400 && gi < 0; attempt++) {
    const gc = 6 + Math.floor(rng() * (GRID_W - 12));
    const gr = GOLDEN_ROW_MIN + Math.floor(rng() * (GOLDEN_ROW_MAX - GOLDEN_ROW_MIN + 1));
    const i = idx(gc, gr);
    if (w.hp[i] > 0 && w.kind[i] !== KIND_BEDROCK) gi = i;
  }
  if (gi < 0) gi = idx(GRID_W >> 1, GOLDEN_ROW_MAX);
  w.kind[gi] = KIND_GOLDEN;
  w.hp[gi] = GOLDEN_HP;
  w.maxHp[gi] = GOLDEN_HP;
  w.goldenIdx = gi;

  // Starting chamber, always clear.
  carve(w, START_COL, START_ROW, START_CHAMBER);
  reveal(w, START_COL, START_ROW, START_CHAMBER + 2.5);

  return w;
}

function carve(w: World, cc: number, cr: number, rad: number) {
  const r = Math.ceil(rad);
  for (let row = cr - r; row <= cr + r; row++) {
    for (let col = cc - r; col <= cc + r; col++) {
      if (!inBounds(col, row)) continue;
      const i = idx(col, row);
      if (w.kind[i] === KIND_BEDROCK || w.kind[i] === KIND_GOLDEN) continue;
      const dc = col - cc;
      const dr = row - cr;
      if (dc * dc + dr * dr <= rad * rad) {
        w.hp[i] = 0;
        w.kind[i] = KIND_ROCK;
      }
    }
  }
}

/** Mark everything within `rad` blocks of (cc, cr) as no longer hidden. */
export function reveal(w: World, cc: number, cr: number, rad: number) {
  const r = Math.ceil(rad);
  const r2 = rad * rad;
  for (let row = cr - r; row <= cr + r; row++) {
    if (row < 0 || row >= GRID_H) continue;
    for (let col = cc - r; col <= cc + r; col++) {
      if (col < 0 || col >= GRID_W) continue;
      const dc = col - cc;
      const dr = row - cr;
      if (dc * dc + dr * dr <= r2) w.seen[idx(col, row)] = 1;
    }
  }
}

// ─── Effects the renderer / UI consumes ───────────────────────────────────────

export type Particle = { x: number; y: number; vx: number; vy: number; life: number; max: number; color: string };
export type Arc = { x1: number; y1: number; x2: number; y2: number; life: number };
export type FloatText = { x: number; y: number; life: number; text: string; color: string };

export type Effects = {
  particles: Particle[];
  arcs: Arc[];
  texts: FloatText[];
  shake: number;
  /** Vibration budget for this frame: 0 none, 1 soft tap, 2 break, 3 token, 4 golden. */
  haptic: number;
  tokensGained: number;
  brokenCount: number;
  goldenHit: boolean;
  goldenBroken: boolean;
};

export function newEffects(): Effects {
  return {
    particles: [], arcs: [], texts: [], shake: 0, haptic: 0,
    tokensGained: 0, brokenCount: 0, goldenHit: false, goldenBroken: false,
  };
}

// ─── Damage ───────────────────────────────────────────────────────────────────

export type DamageCtx = {
  stats: Stats;
  rng: () => number;
  fx: Effects;
  ballLuck: number;
  ballReveal: number;
};

const TIER_COLORS = TIERS.map((t) => t.color);

/**
 * Apply `amount` damage to one block. Returns true if the block broke.
 * `soft` hits (splash, poison, chain) do not themselves trigger haptics.
 */
export function damageBlock(
  w: World, col: number, row: number, amount: number, ctx: DamageCtx, soft: boolean,
): boolean {
  if (!inBounds(col, row)) return false;
  const i = idx(col, row);
  if (w.hp[i] <= 0) return false;
  if (w.kind[i] === KIND_BEDROCK) return false;

  const golden = w.kind[i] === KIND_GOLDEN;
  const next = w.hp[i] - amount;

  const cx = (col + 0.5) * CELL;
  const cy = (row + 0.5) * CELL;

  if (next > 0) {
    w.hp[i] = next;
    if (golden) {
      ctx.fx.goldenHit = true;
      ctx.fx.haptic = Math.max(ctx.fx.haptic, 3);
      spray(ctx.fx, cx, cy, '#ffd23d', 5, ctx.rng);
    } else if (!soft) {
      // A hit that does not break still thumps.
      ctx.fx.haptic = Math.max(ctx.fx.haptic, 1);
      spray(ctx.fx, cx, cy, TIER_COLORS[w.hue[i]], 2, ctx.rng);
    }
    return false;
  }

  // Broken.
  w.hp[i] = 0;
  const kind = w.kind[i];
  w.kind[i] = KIND_ROCK;
  clearPoison(w, i);
  ctx.fx.brokenCount++;
  reveal(w, col, row, ctx.stats.revealRadius + ctx.ballReveal);

  if (kind === KIND_GOLDEN) {
    ctx.fx.goldenBroken = true;
    ctx.fx.haptic = 4;
    ctx.fx.shake = Math.max(ctx.fx.shake, 26);
    spray(ctx.fx, cx, cy, '#ffd23d', 60, ctx.rng);
    return true;
  }

  if (kind === KIND_TOKEN) {
    // Deep rock takes far longer to break, so a token found down there is worth
    // more — otherwise income dries up exactly where upgrades matter most.
    const depthBonus = 1 + tierOfRow(row) * 0.35;
    const gained = Math.max(1, Math.round(ctx.stats.luck * ctx.ballLuck * depthBonus));
    ctx.fx.tokensGained += gained;
    ctx.fx.haptic = Math.max(ctx.fx.haptic, 3);
    ctx.fx.texts.push({ x: cx, y: cy, life: 1.1, text: `+${gained}`, color: '#ffffff' });
    spray(ctx.fx, cx, cy, '#ffffff', 16, ctx.rng);
  } else {
    ctx.fx.haptic = Math.max(ctx.fx.haptic, 2);
    spray(ctx.fx, cx, cy, TIER_COLORS[w.hue[i]], 6, ctx.rng);
    // Luck also shakes the odd token out of plain rock.
    const bonus = (ctx.stats.luck - 1) * ctx.ballLuck * 0.006;
    if (bonus > 0 && ctx.rng() < bonus) {
      ctx.fx.tokensGained += 1;
      ctx.fx.texts.push({ x: cx, y: cy, life: 1.1, text: '+1', color: '#ffffff' });
    }
  }
  ctx.fx.shake = Math.max(ctx.fx.shake, 2.5);
  return true;
}

function spray(fx: Effects, x: number, y: number, color: string, n: number, rng: () => number) {
  if (fx.particles.length > 320) return;
  for (let k = 0; k < n; k++) {
    const a = rng() * Math.PI * 2;
    const sp = 20 + rng() * 110;
    const life = 0.25 + rng() * 0.45;
    fx.particles.push({ x, y, vx: Math.cos(a) * sp, vy: Math.sin(a) * sp, life, max: life, color });
  }
}

function clearPoison(w: World, i: number) {
  if (w.poisonT[i] > 0) {
    w.poisonT[i] = 0;
    w.poisonD[i] = 0;
    w.poisoned.delete(i);
  }
}

function addPoison(w: World, col: number, row: number, dps: number, time: number) {
  if (!inBounds(col, row)) return;
  const i = idx(col, row);
  if (w.hp[i] <= 0 || w.kind[i] === KIND_BEDROCK) return;
  w.poisonD[i] = Math.max(w.poisonD[i], dps);
  w.poisonT[i] = Math.max(w.poisonT[i], time);
  w.poisoned.add(i);
}

// ─── A full impact: direct damage + splash + chain + poison ───────────────────

export function applyImpact(
  w: World, col: number, row: number, ballId: BallId, ctx: DamageCtx,
): boolean {
  const ball = BALLS[ballId];
  const s = ctx.stats;
  const direct = s.damage * s.damageMul * ball.damageMul;

  const broke = damageBlock(w, col, row, direct, ctx, false);

  // Splash.
  const radius = s.splashRadius + ball.splashBonus;
  if (radius > 0 && s.splashFactor + ball.splashBonus * 0.2 > 0) {
    const factor = s.splashFactor + (ball.splashBonus > 0 ? 0.35 : 0);
    const r = Math.ceil(radius);
    for (let dr = -r; dr <= r; dr++) {
      for (let dc = -r; dc <= r; dc++) {
        if (dc === 0 && dr === 0) continue;
        const dist = Math.hypot(dc, dr);
        if (dist > radius) continue;
        const falloff = 1 - dist / (radius + 1);
        damageBlock(w, col + dc, row + dr, direct * factor * falloff, ctx, true);
      }
    }
    ctx.fx.shake = Math.max(ctx.fx.shake, 3 + radius * 1.5);
  }

  // Chain lightning.
  if (ball.chains) {
    chain(w, col, row, ctx, direct);
  }

  // Poison.
  const poisonDps = (ball.poisons ? s.poisonDps : 0) + s.poisonBase;
  if (poisonDps > 0) {
    addPoison(w, col, row, poisonDps, s.poisonTime);
    if (ball.poisons) {
      addPoison(w, col + 1, row, poisonDps * 0.6, s.poisonTime);
      addPoison(w, col - 1, row, poisonDps * 0.6, s.poisonTime);
      addPoison(w, col, row + 1, poisonDps * 0.6, s.poisonTime);
      addPoison(w, col, row - 1, poisonDps * 0.6, s.poisonTime);
    }
  }

  return broke;
}

function chain(w: World, col: number, row: number, ctx: DamageCtx, direct: number) {
  const s = ctx.stats;
  const range = Math.ceil(s.chainRange);
  const targets: number[] = [];
  for (let dr = -range; dr <= range; dr++) {
    for (let dc = -range; dc <= range; dc++) {
      if (dc === 0 && dr === 0) continue;
      const c = col + dc;
      const r = row + dr;
      if (!inBounds(c, r)) continue;
      if (Math.hypot(dc, dr) > s.chainRange) continue;
      const i = idx(c, r);
      if (w.hp[i] > 0 && w.kind[i] !== KIND_BEDROCK) targets.push(i);
    }
  }
  // Shuffle-lite: pick random distinct targets.
  let fromX = (col + 0.5) * CELL;
  let fromY = (row + 0.5) * CELL;
  const n = Math.min(s.chainTargets, targets.length);
  for (let k = 0; k < n; k++) {
    const pick = targets.splice(Math.floor(ctx.rng() * targets.length), 1)[0];
    const tc = pick % GRID_W;
    const tr = (pick / GRID_W) | 0;
    const tx = (tc + 0.5) * CELL;
    const ty = (tr + 0.5) * CELL;
    ctx.fx.arcs.push({ x1: fromX, y1: fromY, x2: tx, y2: ty, life: 0.22 });
    damageBlock(w, tc, tr, direct * s.chainDamage, ctx, true);
    fromX = tx;
    fromY = ty;
  }
}

/** Poison ticks. Called once per frame with the real delta. */
export function tickPoison(w: World, dt: number, ctx: DamageCtx) {
  if (w.poisoned.size === 0) return;
  const spreading = ctx.stats.poisonSpreads;
  const toSpread: number[] = [];
  for (const i of Array.from(w.poisoned)) {
    if (w.hp[i] <= 0) { clearPoison(w, i); continue; }
    const col = i % GRID_W;
    const row = (i / GRID_W) | 0;
    const before = w.hp[i];
    damageBlock(w, col, row, w.poisonD[i] * dt, ctx, true);
    w.poisonT[i] -= dt;
    if (w.hp[i] <= 0 && before > 0 && spreading) toSpread.push(i);
    if (w.poisonT[i] <= 0) clearPoison(w, i);
  }
  for (const i of toSpread) {
    const col = i % GRID_W;
    const row = (i / GRID_W) | 0;
    const dps = ctx.stats.poisonDps * 0.7;
    addPoison(w, col + 1, row, dps, ctx.stats.poisonTime);
    addPoison(w, col - 1, row, dps, ctx.stats.poisonTime);
    addPoison(w, col, row + 1, dps, ctx.stats.poisonTime);
    addPoison(w, col, row - 1, dps, ctx.stats.poisonTime);
  }
}

// ─── Balls ────────────────────────────────────────────────────────────────────

export type Ball = {
  x: number; y: number;
  vx: number; vy: number;
  bounces: number;
  life: number;
  type: BallId;
  /** Recent positions for the dotted trail, newest last: [x, y, x, y, …]. */
  trail: number[];
  dead: boolean;
};

export function spawnVolley(
  px: number, py: number, angle: number, power: number, type: BallId, s: Stats,
): Ball[] {
  const def = BALLS[type];
  const speed = s.speed * def.speedMul * (0.45 + power * 0.55);
  const out: Ball[] = [];
  const n = Math.max(1, Math.round(s.projectiles));
  for (let k = 0; k < n; k++) {
    const offset = n === 1 ? 0 : (k / (n - 1) - 0.5) * s.spreadAngle * (n - 1);
    const a = angle + offset;
    out.push({
      x: px, y: py,
      vx: Math.cos(a) * speed,
      vy: Math.sin(a) * speed,
      bounces: Math.round(s.bounces),
      life: s.lifetime,
      type,
      trail: [px, py],
      dead: false,
    });
  }
  return out;
}

const MAX_STEP = CELL * 0.4;
const TRAIL_POINTS = 14;

export function stepBall(w: World, b: Ball, dt: number, ctx: DamageCtx) {
  const def = BALLS[b.type];
  const speed = Math.hypot(b.vx, b.vy);
  if (speed < 1) { b.dead = true; return; }

  const dist = speed * dt;
  const steps = Math.max(1, Math.ceil(dist / MAX_STEP));
  const h = dt / steps;

  for (let step = 0; step < steps && !b.dead; step++) {
    // Horizontal.
    const nx = b.x + b.vx * h;
    let col = Math.floor(nx / CELL);
    let row = Math.floor(b.y / CELL);
    if (isSolid(w, col, row)) {
      const broke = applyImpact(w, col, row, b.type, ctx);
      if (broke) {
        b.x = nx;
      } else if (def.pierce && !isBedrock(w, col, row)) {
        b.x = nx;
        b.vx *= 0.82;
        b.vy *= 0.82;
      } else {
        b.vx = -b.vx;
        b.bounces--;
      }
    } else {
      b.x = nx;
    }
    if (b.bounces < 0) { b.dead = true; break; }

    // Vertical.
    const ny = b.y + b.vy * h;
    col = Math.floor(b.x / CELL);
    row = Math.floor(ny / CELL);
    if (isSolid(w, col, row)) {
      const broke = applyImpact(w, col, row, b.type, ctx);
      if (broke) {
        b.y = ny;
      } else if (def.pierce && !isBedrock(w, col, row)) {
        b.y = ny;
        b.vx *= 0.82;
        b.vy *= 0.82;
      } else {
        b.vy = -b.vy;
        b.bounces--;
      }
    } else {
      b.y = ny;
    }
    if (b.bounces < 0) { b.dead = true; break; }
  }

  b.life -= dt;
  if (b.life <= 0) b.dead = true;

  const bc = Math.floor(b.x / CELL);
  const br = Math.floor(b.y / CELL);
  reveal(w, bc, br, ctx.stats.revealRadius + def.revealBonus);

  b.trail.push(b.x, b.y);
  if (b.trail.length > TRAIL_POINTS * 2) b.trail.splice(0, b.trail.length - TRAIL_POINTS * 2);
}

function isBedrock(w: World, col: number, row: number): boolean {
  if (!inBounds(col, row)) return true;
  return w.kind[idx(col, row)] === KIND_BEDROCK;
}

// ─── Player movement ──────────────────────────────────────────────────────────

export const PLAYER_REACH = 15;

/**
 * Walk from the player toward a tapped block, stopping just before the first
 * wall. Returns the block the player ends up in, or null if it cannot move.
 */
export function walkToward(
  w: World, fromCol: number, fromRow: number, toCol: number, toRow: number,
): { col: number; row: number } | null {
  const dc = toCol - fromCol;
  const dr = toRow - fromRow;
  const dist = Math.hypot(dc, dr);
  if (dist < 0.5) return null;

  const capped = Math.min(dist, PLAYER_REACH);
  const stepC = dc / dist;
  const stepR = dr / dist;

  let lastCol = fromCol;
  let lastRow = fromRow;
  const samples = Math.ceil(capped * 4);
  for (let k = 1; k <= samples; k++) {
    const t = (k / samples) * capped;
    const c = Math.round(fromCol + stepC * t);
    const r = Math.round(fromRow + stepR * t);
    if (!inBounds(c, r) || w.hp[idx(c, r)] > 0) break;
    lastCol = c;
    lastRow = r;
  }
  if (lastCol === fromCol && lastRow === fromRow) return null;
  return { col: lastCol, row: lastRow };
}

// ─── Save / load ──────────────────────────────────────────────────────────────

export type SaveData = {
  v: number;
  seed: number;
  tokens: number;
  unlocked: string[];
  ball: BallId;
  col: number;
  row: number;
  deepest: number;
  broken: number;
  won: boolean;
  /** Run-length encoded hp grid: [count, value, count, value, …]. */
  hp: number[];
  /** Run-length encoded seen grid. */
  seen: number[];
};

export function rle(arr: ArrayLike<number>): number[] {
  const out: number[] = [];
  let run = 1;
  for (let i = 1; i <= arr.length; i++) {
    if (i < arr.length && arr[i] === arr[i - 1] && run < 65535) {
      run++;
    } else {
      out.push(run, arr[i - 1]);
      run = 1;
    }
  }
  return out;
}

export function unrle(data: number[], into: Int16Array | Uint8Array): boolean {
  let p = 0;
  for (let i = 0; i + 1 < data.length; i += 2) {
    const count = data[i];
    const value = data[i + 1];
    for (let k = 0; k < count; k++) {
      if (p >= into.length) return false;
      into[p++] = value;
    }
  }
  return p === into.length;
}

File: roblox/Engine.lua
--[[ BlindShot Game Engine - Roblox Version
Place this as a ModuleScript in: StarterPlayer > StarterPlayerScripts > Engine
]]

local Config = require(script.Parent:WaitForChild("Config"))
local Engine = {}

local function noise2D(x, y, seed)
	local n = math.sin(x * 12.9898 + y * 78.233 + seed) * 43758.5453
	return n - math.floor(n)
end

function Engine.genWorld(seed)
	local world = {
		seed = seed,
		spawnCol = Config.START_COL,
		spawnRow = Config.START_ROW,
		hp = {},
		hue = {},
		kind = {},
		shade = {},
		seen = {},
		poisonT = {},
		maxHp = {},
		goldenIdx = -1,
		goldenSighted = false,
	}

	for i = 1, Config.CELLS do
		world.hp[i] = 0
		world.hue[i] = 1
		world.kind[i] = Config.KIND_SOLID
		world.shade[i] = 128
		world.seen[i] = 0
		world.poisonT[i] = 0
		world.maxHp[i] = 0
	end

	local rng = seed
	local function nextRng()
		rng = (rng * 1103515245 + 12345) % (2^31)
		return rng / (2^31)
	end

	for row = 0, Config.GRID_H - 1 do
		for col = 0, Config.GRID_W - 1 do
			local i = row * Config.GRID_W + col

			local dc = col - Config.START_COL
			local dr = row - Config.START_ROW
			local dist = math.sqrt(dc * dc + dr * dr)

			if col == 0 or col == Config.GRID_W - 1 or row == 0 or row == Config.GRID_H - 1 then
				world.kind[i] = Config.KIND_BEDROCK
				world.hp[i] = 9999
				world.maxHp[i] = 9999
				world.hue[i] = 0
				goto continue
			end

			if dist < 8 then
				world.hp[i] = 0
				world.seen[i] = 1
				goto continue
			end

			if dist > 250 then
				world.hp[i] = 0
				goto continue
			end

			local n = noise2D(col * 0.1, row * 0.1, seed)
			local threshold = 0.5 + dist / 500

			if n < threshold then
				world.hp[i] = 0
				goto continue
			end

			local baseHp = Config.hpAtDist(dist)
			world.hp[i] = math.max(1, math.floor(baseHp * (0.75 + nextRng() * 0.55)))
			world.maxHp[i] = world.hp[i]
			world.hue[i] = (math.floor(dist / 8) % #Config.BLOCKS) + 1
			world.shade[i] = 128 + math.floor((nextRng() - 0.5) * 80)

			if nextRng() < (1 / Config.TOKEN_RARITY) then
				world.kind[i] = Config.KIND_TOKEN
				world.hp[i] = 1
				world.maxHp[i] = 1
			end

			if dist >= Config.GOLDEN_DIST_MIN and dist <= Config.GOLDEN_DIST_MAX and
			   world.goldenIdx == -1 and nextRng() < 0.001 then
				world.kind[i] = Config.KIND_GOLDEN
				world.hp[i] = math.max(30, math.floor(Config.hpAtDist(dist) * 3))
				world.maxHp[i] = world.hp[i]
				world.goldenIdx = i
			end

			::continue::
		end
	end

	return world
end

function Engine.distFrom(c1, r1, c2, r2)
	local dc = c1 - c2
	local dr = r1 - r2
	return math.sqrt(dc * dc + dr * dr)
end

function Engine.blockedAt(world, x, y, radius)
	local l = math.floor((x - radius) / Config.CELL)
	local r = math.floor((x + radius) / Config.CELL)
	local t = math.floor((y - radius) / Config.CELL)
	local b = math.floor((y + radius) / Config.CELL)

	local function solidAt(c, row)
		if c < 0 or c >= Config.GRID_W or row < 0 or row >= Config.GRID_H then
			return true
		end
		local i = row * Config.GRID_W + c
		return world.hp[i] > 0
	end

	return solidAt(l, t) or solidAt(r, t) or solidAt(l, b) or solidAt(r, b)
end

function Engine.stepPlayer(player, world, dt)
	local tvx = player.ix * Config.MOVE_SPEED
	local tvy = player.iy * Config.MOVE_SPEED
	local moving = (player.ix ~= 0 or player.iy ~= 0)
	local k = 1 - math.exp(-(moving and Config.MOVE_ACCEL or Config.MOVE_DECEL) * dt)

	player.vx = player.vx + (tvx - player.vx) * k
	player.vy = player.vy + (tvy - player.vy) * k

	if not moving and math.sqrt(player.vx * player.vx + player.vy * player.vy) < 1 then
		player.vx = 0
		player.vy = 0
	end

	if player.vx == 0 and player.vy == 0 then return end

	local nx = player.x + player.vx * dt
	if not Engine.blockedAt(world, nx, player.y, Config.PLAYER_RADIUS) then
		player.x = nx
	else
		local col = math.floor((nx + (player.vx > 0 and 1 or -1) * Config.PLAYER_RADIUS) / Config.CELL)
		player.x = player.vx > 0 and col * Config.CELL - Config.PLAYER_RADIUS - 0.01 or (col + 1) * Config.CELL + Config.PLAYER_RADIUS + 0.01
		player.vx = 0
	end

	local ny = player.y + player.vy * dt
	if not Engine.blockedAt(world, player.x, ny, Config.PLAYER_RADIUS) then
		player.y = ny
	else
		local row = math.floor((ny + (player.vy > 0 and 1 or -1) * Config.PLAYER_RADIUS) / Config.CELL)
		player.y = player.vy > 0 and row * Config.CELL - Config.PLAYER_RADIUS - 0.01 or (row + 1) * Config.CELL + Config.PLAYER_RADIUS + 0.01
		player.vy = 0
	end

	player.col = math.floor(player.x / Config.CELL)
	player.row = math.floor(player.y / Config.CELL)
end

function Engine.breakBlock(world, col, row)
	if col < 0 or col >= Config.GRID_W or row < 0 or row >= Config.GRID_H then
		return 0
	end
	local i = row * Config.GRID_W + col
	if world.hp[i] <= 0 then return 0 end

	world.hp[i] = 0
	world.seen[i] = 1
	return 1
end

function Engine.damageBlock(world, col, row, damage)
	if col < 0 or col >= Config.GRID_W or row < 0 or row >= Config.GRID_H then
		return 0
	end
	local i = row * Config.GRID_W + col
	if world.hp[i] <= 0 then return 0 end

	world.hp[i] = world.hp[i] - damage
	if world.hp[i] <= 0 then
		world.hp[i] = 0
		world.seen[i] = 1
		return 1
	end
	return 0
end

function Engine.newLightMap()
	local lm = {
		r = {},
		g = {},
		b = {},
	}
	for i = 1, Config.CELLS do
		lm.r[i] = Config.LIGHT_AMBIENT
		lm.g[i] = Config.LIGHT_AMBIENT
		lm.b[i] = Config.LIGHT_AMBIENT
	end
	return lm
end

function Engine.buildLightMap(world, lightMap, ballList)
	for i = 1, Config.CELLS do
		lightMap.r[i] = Config.LIGHT_AMBIENT
		lightMap.g[i] = Config.LIGHT_AMBIENT
		lightMap.b[i] = Config.LIGHT_AMBIENT
	end

	for _, ball in ipairs(ballList) do
		local bc = math.floor(ball.x / Config.CELL)
		local br = math.floor(ball.y / Config.CELL)
		for r = br - Config.LIGHT_RANGE, br + Config.LIGHT_RANGE do
			for c = bc - Config.LIGHT_RANGE, bc + Config.LIGHT_RANGE do
				if c >= 0 and c < Config.GRID_W and r >= 0 and r < Config.GRID_H then
					local i = r * Config.GRID_W + c
					local dc = c - bc
					local dr = r - br
					local dist = math.sqrt(dc * dc + dr * dr)
					local falloff = math.exp(-Config.LIGHT_FALLOFF * dist)
					lightMap.r[i] = lightMap.r[i] + falloff * 0.3
					lightMap.g[i] = lightMap.g[i] + falloff * 0.3
					lightMap.b[i] = lightMap.b[i] + falloff * 0.3
				end
			end
		end
	end

	if world.goldenIdx >= 0 and world.hp[world.goldenIdx] > 0 then
		local c = world.goldenIdx % Config.GRID_W
		local r = math.floor(world.goldenIdx / Config.GRID_W)
		for r2 = r - Config.GOLDEN_AURA, r + Config.GOLDEN_AURA do
			for c2 = c - Config.GOLDEN_AURA, c + Config.GOLDEN_AURA do
				if c2 >= 0 and c2 < Config.GRID_W and r2 >= 0 and r2 < Config.GRID_H then
					local i = r2 * Config.GRID_W + c2
					local dc = c2 - c
					local dr = r2 - r
					local dist = math.sqrt(dc * dc + dr * dr)
					local falloff = math.exp(-0.08 * dist)
					lightMap.r[i] = lightMap.r[i] + falloff * 0.8
					lightMap.g[i] = lightMap.g[i] + falloff * 0.6
					lightMap.b[i] = lightMap.b[i] + falloff * 0.2
				end
			end
		end
	end
end

return Engine
